sklearn/cluster/optics_.py

```diff
@@ -250,469 +218,322 @@ def fit(self, X, y=None):
-        # Neighbors of current point are already processed.
-        if not unproc.size:
-            return
-
-        # Only compute distances to unprocessed neighbors:
-        if self.metric == 'precomputed':
-            dists = X[point_index, unproc]
-        else:
-            dists = pairwise_distances(P, np.take(X, unproc, axis=0),
-                                       self.metric, n_jobs=None).ravel()
-
-        rdists = np.maximum(dists, self.core_distances_[point_index])
-        improved = np.where(rdists < np.take(self.reachability_, unproc))
-        self.reachability_[unproc[improved]] = rdists[improved]
-        self.predecessor_[unproc[improved]] = point_index
-
-    def extract_dbscan(self, eps):
-        """Performs DBSCAN extraction for an arbitrary epsilon.
-
-        Extraction runs in linear time. Note that if the `max_eps` OPTICS
-        parameter was set to < inf for extracting reachability and ordering
-        arrays, DBSCAN extractions will be unstable for `eps` values close to
-        `max_eps`. Setting `eps` < (`max_eps` / 5.0) will guarantee
-        extraction parity with DBSCAN.
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        The data.
+    neighbors : NearestNeighbors instance
+        The fitted nearest neighbors estimator.
+    working_memory : int, optional
+        The sought maximum memory for temporary distance matrix chunks.
+        When None (default), the value of
+        ``sklearn.get_config()['working_memory']`` is used.
 
-        Parameters
-        ----------
-        eps : float or int, required
-            DBSCAN `eps` parameter. Must be set to < `max_eps`. Equivalence
-            with DBSCAN algorithm is achieved if `eps` is < (`max_eps` / 5)
+    Returns
+    -------
+    core_distances : array, shape (n_samples,)
+        Distance at which each sample becomes a core point.
+        Points which will never be core have a distance of inf.
+    """
+    n_samples = len(X)
+    core_distances = np.empty(n_samples)
+    core_distances.fill(np.nan)
 
-        Returns
-        -------
-        core_sample_indices_ : array, shape (n_core_samples,)
-            The indices of the core samples.
+    chunk_n_rows = get_chunk_n_rows(row_bytes=16 * min_samples,
+                                    max_n_rows=n_samples,
+                                    working_memory=working_memory)
+    slices = gen_batches(n_samples, chunk_n_rows)
+    for sl in slices:
+        core_distances[sl] = neighbors.kneighbors(
+            X[sl], min_samples)[0][:, -1]
+    return core_distances
 
-        labels_ : array, shape (n_samples,)
-            The estimated labels.
-        """
-        check_is_fitted(self, 'reachability_')
 
-        if eps > self.max_eps:
-            raise ValueError('Specify an epsilon smaller than %s. Got %s.'
-                             % (self.max_eps, eps))
+def compute_optics_graph(X, min_samples, max_eps, metric, p, metric_params,
```

Comment:
A crazy thought: How about calling this function optics?

Reply:
we talked about it with Joel and Gael, optics kinda implies that it also does the labling of the data, which it doesn't. So we decided to be explicit.

Reply:
that's fine, thanks